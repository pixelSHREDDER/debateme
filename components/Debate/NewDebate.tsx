'use client'

import { addDebate } from '@/actions/add-debate';
import { useFormState } from 'react-dom'

const initialState = {
  message: '',
}

export default function NewDebate({ creatorId }: { creatorId: number }) {
  const [state, formAction] = useFormState(addDebate, initialState)

  return (
    <form action={formAction}>
      <input
        type="hidden"
        id="creatorId"
        name="creatorId"
        value={creatorId} />
      <label htmlFor="topic">Topic</label>
      <input type="text" id="topic" name="topic" required />
      <p aria-live="polite" className="sr-only">
        {state?.message}
      </p>
      <button type="submit">Start Debating!</button>
    </form>
  )
}