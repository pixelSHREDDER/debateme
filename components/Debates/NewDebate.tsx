'use client'

import addDebate from '@/actions/add-debate'
import { useUser } from '@auth0/nextjs-auth0/client'
import { useFormState } from 'react-dom'
import FormSubmitButton from '../Forms/FormSubmitButton'

const initialState = {
  message: '',
}

export default function NewDebate() {
  const [state, formAction] = useFormState(addDebate, initialState)
  const { user } = useUser()

  return user && user.sub ?
    <form action={formAction}>
      <input
        type="hidden"
        id="creatorSub"
        name="creatorSub"
        value={user.sub} />
      <label htmlFor="topic">Topic</label>
      <input type="text" id="topic" name="topic" required />
      <label htmlFor="cooldownMins">Cooldown period between turns</label>
      <select id="cooldownMins" name="cooldownMins" defaultValue="60" required>
        <option value="180">3 hours</option>
        <option value="120">2 hours</option>
        <option value="60">1 hour</option>
        <option value="10">10 minutes</option>
        <option value="1">1 minute</option>
        <option value="0">No cooldown</option>
      </select>
      <p aria-live="polite">
        {state?.message}
      </p>
      <FormSubmitButton label="Start Debating!" />
    </form> :
  'login'
}
