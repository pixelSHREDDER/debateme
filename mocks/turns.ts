import { TTurn } from '@/lib/prisma-types'

// Lorem Ipsum HTML courtesy of https://css-tricks.com/new-site-html-ipsum/

export const createTurns: () => TTurn[] = () => {
  const createdAt = new Date('2024-04-13T12:24:02Z')

  return [
    {
      id: 1,
      body: `
          <p>Didn't melt fairer keepsakes since Fellowship elsewhere.</p>
          <p>Woodlands payment <a href="#">Osgiliath</a> tightening. Barad-dur follow belly comforts tender tough bell? Many that live deserve death. Some that die deserve life. Outwitted teatime grasp defeated before stones reflection corset seen animals Saruman's call?</p>
          <p>Curse you and all the halflings! Deserted anytime Lake-town burned caves balls. Smoked lthilien forbids Thrain? <span style="font-size: 2.0rem">Tad survive ensnare joy mistake courtesy Bagshot Row.</span> Ligulas step drops both? You shall not pass!</p>
          <p>Branch hey-diddle-diddle pony trouble'll sleeping during jump Narsil. <span style="font-size: 0.6rem">North valor overflowing sort Iáve mister kingly money?</span> Tender respectable success Valar impressive unfriendly bloom scraped?</p>
          <ul>
            <li>Adamant.</li>
            <li>Southfarthing!</li>
            <li>Witch-king.</li>
            <li>Precious.</li>
            <li>Gaffer's!</li>
          </ul>
          <ul>
            <li>Excuse tightening yet survives two cover Undómiel city ablaze.</li>
            <li>Keepsakes deeper clouds Buckland position 21 lied bicker fountains ashamed.</li>
            <li>Women rippling cold steps rules Thengel finer.</li>
            <li>Portents close Havens endured irons hundreds handle refused sister?</li>
            <li>Harbor Grubbs fellas riddles afar!</li>
          </ul>
          <p>Ravens wonder wanted runs me crawl gaining lots faster! Khazad-dum surprise baby season ranks. I bid you all a very fond farewell.</p>
          <ol>
            <li>Narsil.</li>
            <li>Elros.</li>
            <li>Arwen Evenstar.</li>
            <li>Maggot's?</li>
            <li>Bagginses?</li>
          </ol>
          <ol>
            <li>Concerning Hobbits l golf air fifth bell prolonging camp.</li>
            <li>Grond humble rods nearest mangler.</li>
            <li>Enormity Lórien merry gravy stayed move.</li>
            <li>Diversion almost notion furs between fierce laboring Nazgûl ceaselessly parent.</li>
            <li>Agree ruling um wasteland Bagshot Row expect sleep.</li>
          </ol>
          <p>Delay freezes Gollum. Let the Ring-bearer decide. Bagshot Row chokes pole pauses immediately orders taught éored musing three-day? Disease rune repel source fire Goblinses already?</p>
          <hr>
          <p>What about second breakfast? Nags runt near Lindir lock discover level? Andûril breathe waited flatten union.</p>
          <blockquote>
            <p>You shall be the Fellowship of the Ring.</p>
            <footer>—Númenor, <cite>sweeter burned verse</cite></footer>
          </blockquote>
          <p>I think we should get off the road. Penalty sight splintered Misty Mountain mithril? Unrest lasts rode league bears absence Bracegirdle athletic contract nice parent slowed?</p>
          <pre>Pardon Concerning Hobbits rune goblins? Twitching figure including rightful Thorin's level! Worth tubers threats Hornburg deadliest? Unfold thumping shh wants Homely!</pre>
          <hr>`,
      debateId: 1,
      createdAt,
      userSub: 'auth0|1234567890abcdefghij1234',
      },
      {
        id: 2,
        body: `
          <p>Didn't melt fairer keepsakes since Fellowship elsewhere.</p>
          <p>Woodlands payment <a href="#">Osgiliath</a> tightening. Barad-dur follow belly comforts tender tough bell? Many that live deserve death. Some that die deserve life. Outwitted teatime grasp defeated before stones reflection corset seen animals Saruman's call?</p>
          <p><span style="font-size: 2.0rem">Tad survive ensnare joy mistake courtesy Bagshot Row.</span></p>
          <p>Ligulas step drops both? You shall not pass! Tender respectable success Valar impressive unfriendly bloom scraped? Branch hey-diddle-diddle pony trouble'll sleeping during jump Narsil.</p>
          <p><span style="font-size: 0.6rem">North valor overflowing sort Iáve mister kingly money?</span></p>
          <p>Curse you and all the halflings! Deserted anytime Lake-town burned caves balls. Smoked lthilien forbids Thrain?</p>
          <ul>
            <li>Adamant.</li>
            <li>Southfarthing!</li>
            <li>Witch-king.</li>
            <li>Precious.</li>
            <li>Gaffer's!</li>
          </ul>
          <ul>
            <li>Excuse tightening yet survives two cover Undómiel city ablaze.</li>
            <li>Keepsakes deeper clouds Buckland position 21 lied bicker fountains ashamed.</li>
            <li>Women rippling cold steps rules Thengel finer.</li>
            <li>Portents close Havens endured irons hundreds handle refused sister?</li>
            <li>Harbor Grubbs fellas riddles afar!</li>
          </ul>
          <p>Ravens wonder wanted runs me crawl gaining lots faster! Khazad-dum surprise baby season ranks. I bid you all a very fond farewell.</p>
          <ol>
            <li>Narsil.</li>
            <li>Elros.</li>
            <li>Arwen Evenstar.</li>
            <li>Maggot's?</li>
            <li>Bagginses?</li>
          </ol>
          <ol>
            <li>Concerning Hobbits l golf air fifth bell prolonging camp.</li>
            <li>Grond humble rods nearest mangler.</li>
            <li>Enormity Lórien merry gravy stayed move.</li>
            <li>Diversion almost notion furs between fierce laboring Nazgûl ceaselessly parent.</li>
            <li>Agree ruling um wasteland Bagshot Row expect sleep.</li>
          </ol>
          <p>Delay freezes Gollum. Let the Ring-bearer decide. Bagshot Row chokes pole pauses immediately orders taught éored musing three-day? Disease rune repel source fire Goblinses already?</p>
          <hr>
          <p>What about second breakfast? Nags runt near Lindir lock discover level? Andûril breathe waited flatten union.</p>
          <blockquote>
            <p>You shall be the Fellowship of the Ring.</p>
            <footer>—Númenor, <cite>sweeter burned verse</cite></footer>
          </blockquote>
          <p>I think we should get off the road. Penalty sight splintered Misty Mountain mithril? Unrest lasts rode league bears absence Bracegirdle athletic contract nice parent slowed?</p>
          <pre>Pardon Concerning Hobbits rune goblins? Twitching figure including rightful Thorin's level! Worth tubers threats Hornburg deadliest? Unfold thumping shh wants Homely!</pre>`,
        debateId: 1,
        createdAt,
        userSub: 'auth0|abcdefghij1234567890abcd',
        },
    ] as const
  }
