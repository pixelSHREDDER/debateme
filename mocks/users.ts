import { UserProfile } from '@auth0/nextjs-auth0/client'

export const getUsers: () => UserProfile[] = () => [
    {
      nickname: 'ron.swanson12345',
      name: 'Ron Swanson',
      picture: 'https://s.gravatar.com/avatar/7642279092c16299a6a93e7523cc9ed2?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Frs.png',
      updated_at: '2024-08-17T04:12:58.145Z',
      email: 'ron.swanson12345@excellent-parser.name',
      email_verified: false,
      sub: 'auth0|1234567890abcdefghij1234',
      sid: 'o_abcdefgh12345678-ABCDEF12345',
    },
    {
      nickname: 'leslie.knope54321',
      name: 'Leslie Knope',
      picture: 'https://s.gravatar.com/avatar/7642279092c16299a6a93e7523cc9ed2?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Flk.png',
      updated_at: '2024-08-17T04:12:58.145Z',
      email: 'leslie.knope54321@excellent-parser.name',
      email_verified: false,
      sub: 'auth0|abcdefghij1234567890abcd',
      sid: 'o_12345678abcdefgh-12345ABCDEF',
    },
  ] as const
