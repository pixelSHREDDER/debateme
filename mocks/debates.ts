import { TDebate } from '@/lib/prisma-types'
import { DebateStatus } from '@prisma/client'
import { createTurns } from './turns'

export const createDebates: () => TDebate[] = () => {
  const createdAtDate = new Date('2024-04-13T12:24:02Z')
  const cooldownDate = new Date()
  const createdTurns = createTurns()

  return [
    {
      id: 1,
      topic: 'Your turn',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      turn: createdTurns,
      status: DebateStatus.CreatorTurn,
    },
    {
      id: 2,
      topic: 'Opponent\'s Turn',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      status: DebateStatus.OpponentTurn,
      turn: [
        createdTurns[0],
      ],
    },
    {
      id: 3,
      topic: 'Wait for Cooldown',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: 'auth0|abcdefghij1234567890abcd',
      cooldownMins: 60,
      status: DebateStatus.CreatorCooldown,
      turn: [
        createdTurns[0],
        {
          ...createdTurns[1],
          createdAt: cooldownDate,
        },
      ],
    },
    {
      id: 4,
      topic: 'No opponent',
      createdAt: createdAtDate,
      creatorSub: 'auth0|1234567890abcdefghij1234',
      opponentSub: null,
      cooldownMins: 60,
      status: DebateStatus.NoOpponent,
    },
    {
      id: 5,
      topic: 'You\'re Invited to join',
      createdAt: createdAtDate,
      creatorSub: 'auth0|abcdefghij1234567890abcd',
      opponentSub: null,
      cooldownMins: 60,
      status: DebateStatus.NoOpponent,
    },
  ] as const
}
