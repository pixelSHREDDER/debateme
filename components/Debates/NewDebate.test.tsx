import * as auth0Hooks from '@auth0/nextjs-auth0/client'
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import { fireEvent, render, screen } from '@testing-library/react'
import ProviderWrapper from '@/tests/helpers/ProviderWrapper'
import { getUsers } from '@/mocks/users'
import NewDebate from './NewDebate'

const users: auth0Hooks.UserProfile[] = getUsers()

describe('New Debate component', () => {
  const useUserSpy = vi.spyOn(auth0Hooks, 'useUser')

  beforeEach(() => {
    vi.useFakeTimers()
    useUserSpy.mockReturnValue({
      checkSession: vi.fn(),
      isLoading: false,
      user: users[0],
    })
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  describe('New Debate', () => {
    test('renders correctly', () => {
      render(
        <NewDebate />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.getByTestId('new-debate-form')).toMatchSnapshot()
    })

    test('does not render if user is loading', () => {
      useUserSpy.mockReturnValueOnce({
        checkSession: vi.fn(),
        isLoading: true,
        user: users[0],
      })
      render(
        <NewDebate />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.queryByTestId('new-debate-form')).toBeNull
    })

    test('does not render if user is not defined', () => {
      useUserSpy.mockReturnValueOnce({
        checkSession: vi.fn(),
        isLoading: false,
      })
      render(
        <NewDebate />,
        { wrapper: ProviderWrapper },
      )
      expect(screen.queryByTestId('new-debate-form')).toBeNull
    })
  })

  describe('Submitting new debate', () => {
    test('submits new debate when button is clicked', async () => {
      render(
        <NewDebate />,
        { wrapper: ProviderWrapper },
      )
      fireEvent.submit(screen.getByTestId('new-debate-form'))
    })
  })
})
