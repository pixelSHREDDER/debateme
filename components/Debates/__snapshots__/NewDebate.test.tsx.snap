// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`New Debate component > New Debate > renders correctly 1`] = `
<form
  data-testid="new-debate-form"
>
  <fieldset
    class="m_ef274e49 m_e9408a47 mantine-Fieldset-root"
    data-variant="filled"
    radius="lg"
    style="--fieldset-radius: var(--mantine-radius-lg); margin-inline: 0rem;"
  >
    <div
      class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
    >
      <div
        class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
        data-variant="default"
      >
        <input
          aria-invalid="false"
          class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
          data-variant="default"
          id="creatorSub"
          name="creatorSub"
          type="hidden"
          value="auth0|1234567890abcdefghij1234"
        />
      </div>
    </div>
    <div
      class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
    >
      <label
        class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
        data-required="true"
        for="topic"
        id="topic-label"
      >
        Topic
        <span
          aria-hidden="true"
          class="m_78a94662 mantine-InputWrapper-required mantine-TextInput-required"
        >
           *
        </span>
      </label>
      <div
        class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
        data-variant="default"
      >
        <input
          aria-invalid="false"
          class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
          data-variant="default"
          id="topic"
          name="topic"
          required=""
          type="text"
        />
      </div>
    </div>
    <div
      class="m_46b77525 mantine-InputWrapper-root mantine-NativeSelect-root"
      style="max-width: max(50%, 200px);"
    >
      <label
        class="m_8fdc1311 mantine-InputWrapper-label mantine-NativeSelect-label"
        data-required="true"
        for="cooldownMins"
        id="cooldownMins-label"
      >
        Cooldown period between turns
        <span
          aria-hidden="true"
          class="m_78a94662 mantine-InputWrapper-required mantine-NativeSelect-required"
        >
           *
        </span>
      </label>
      <p
        class="m_fe47ce59 mantine-InputWrapper-description mantine-NativeSelect-description"
        id="cooldownMins-description"
      >
        How long will you need to respond to what your opponent says, rather than react to it?
      </p>
      <div
        class="m_6c018570 mantine-Input-wrapper mantine-NativeSelect-wrapper"
        data-pointer="true"
        data-variant="default"
        data-with-right-section="true"
        style="--input-margin-top: calc(var(--mantine-spacing-xs) / 2); --input-right-section-pointer-events: none;"
      >
        <select
          aria-describedby="cooldownMins-description"
          aria-invalid="false"
          class="m_8fb7ebe7 mantine-Input-input mantine-NativeSelect-input"
          data-variant="default"
          id="cooldownMins"
          name="cooldownMins"
          required=""
        >
          <option
            value="180"
          >
            3 hours
          </option>
          <option
            value="120"
          >
            2 hours
          </option>
          <option
            value="60"
          >
            1 hour
          </option>
          <option
            value="10"
          >
            10 minutes
          </option>
          <option
            selected=""
            value="5"
          >
            5 minutes
          </option>
          <option
            value="1"
          >
            1 minute
          </option>
          <option
            value="0"
          >
            No cooldown
          </option>
        </select>
        <div
          class="m_82577fc2 mantine-Input-section mantine-NativeSelect-section"
          data-position="right"
        >
          <svg
            class="m_2943220b mantine-ComboboxChevron-chevron"
            data-combobox-chevron="true"
            fill="none"
            viewBox="0 0 15 15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="m_8bffd616 mantine-Flex-root __m__-rh"
      style="gap: calc(1.25rem * var(--mantine-scale)); align-items: center; justify-content: flex-start;"
    >
      <aside
        aria-describedby="alert_quotes_:rj:-body"
        class="_quotes_e6219c m_66836ed3 mantine-Alert-root"
        data-variant="outline"
        id="alert_quotes_:rj:"
        role="alert"
        style="--alert-radius: var(--mantine-radius-md); --alert-bg: transparent; --alert-color: #9562d2; --alert-bd: calc(0.0625rem * var(--mantine-scale)) solid #9562d2; font-style: italic; width: auto; max-width: 75%;"
      >
        <div
          class="m_a5d60502 mantine-Alert-wrapper"
        >
          <div
            class="m_667c2793 mantine-Alert-body"
          >
            <div
              class="m_7fa78076 mantine-Alert-message"
              data-variant="outline"
              id="alert_quotes_:rj:-body"
            >
              “Don’t be a dick.” ― Unknown
            </div>
          </div>
        </div>
      </aside>
      <button
        class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
        data-variant="filled"
        style="--button-radius: var(--mantine-radius-md); --button-bg: var(--mantine-color-marble-filled); --button-hover: var(--mantine-color-marble-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; flex-shrink: 0;"
        type="submit"
      >
        <span
          class="m_80f1301b mantine-Button-inner"
        >
          <span
            class="m_811560b9 mantine-Button-label"
          >
            Start Debating!
          </span>
        </span>
      </button>
    </div>
  </fieldset>
</form>
`;
